### DISCUSSION TITLE
How to add a model to the swagger file?

### DISCUSSION THREAD

**Discussion by Ghostbird** [domaindrivendev/Swashbuckle.AspNetCore#3061] [Q&A] [answered]
I have a middleware that catches and packages exceptions that occur on our API in a common `ApiException` class. That class is mentioned nowhere on any controllers. I want it included in the `swagger.json`. I read [this stackoverflow](https://stackoverflow.com/questions/49006079/using-swashbuckle-for-asp-net-core-how-can-i-add-a-model-to-the-generated-model) but the code examples given there are for older versions of `Swashbuckle` and neither works with `Swashbuckle` 6+.

This is the code I tried:

```csharp
  sealed class GenerateSchemaFor<T> : IDocumentFilter where T : class
  {
    public void Apply(OpenApiDocument swaggerDoc, DocumentFilterContext context)
    {
      if (!context.SchemaRepository.Schemas.ContainsKey(typeof(T).Name))
      {
        context.SchemaGenerator.GenerateSchema(typeof(T), context.SchemaRepository);
      }
      if (!context.SchemaRepository.Schemas.ContainsKey(typeof(T).Name))
      {
        throw new SwaggerGeneratorException($"Could not register schema for {typeof(T).Name}!");
      }
    }
  }
```

I call it in the swagger configuration:

```csharp
  .AddSwaggerGen(options => options.AddModel<ApiException>())`
```

However this always throws: `Could not register schema for ApiException!`

It's quite odd. I've read through the code of `SchemaGenerator.cs` and it seems it ought to work. Yet it doesn't.

**Comment by martincostello** [comment]
We'd need to see more information (an inner exception and a stack trace) to determine whether `Could not register schema for ApiException!` is a bug or not.

However you should be able to use [filters](https://github.com/domaindrivendev/Swashbuckle.AspNetCore?tab=readme-ov-file#extend-generator-with-operation-schema--document-filters) to add an additional error response to all of your operations to handle the behaviour your middleware adds.

The code for the filter you've provided looks wrong as you create a schema, but then you don't actually add it to the document.

**Reply by Ghostbird** [reply]
As you can see from my provided code `Could not register schema for ApiException!` is my own error I throw when I have concluded that the schema was not generated and/or added correctly.

Why do you think that I create the schema but don't add it?

I went by the examples from [Stack Overflow](https://stackoverflow.com/questions/49006079/using-swashbuckle-for-asp-net-core-how-can-i-add-a-model-to-the-generated-model) and [this test](test/Swashbuckle.AspNetCore.SwaggerGen.Test/Fixtures/TestDocumentFilter.cs). When I call `GenerateSchema` the schema is auto-added to the passed `SchemaRepository` in [this line]( https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/5b63e97ca0829dd2a9cbc27ccba670f399e37540/src/Swashbuckle.AspNetCore.SwaggerGen/SchemaGenerator/SchemaGenerator.cs#L530). None of the examples I've seen actually add it to the document in addition to adding it to the schema repository.

The full call stack that adds the schema to the repository would be:

- `GenerateSchema`
- `GenerateSchemaForType`
- `GenerateConcreteSchema` because `dataContract.DataType` is `DataType.Object` which sets `returnAsReference` it then calls
- `GenerateReferencedSchema` which calls
- `schemaRepository.AddDefinition(schemaId, schema)`

**Reply by martincostello** [reply]
> Why do you think that I create the schema but don't add it?

I assumed that because you weren't using the result of the `GenerateSchema()` call it wasn't added to the `OpenApiDocument`.

**Reply by Ghostbird** [reply]
Well, I though it was a bit strange too, but the examples that I saw never add the result to the `OpenApiDocument`. Apparently that's not necessary? It _is_ added to the `context.SchemaRepository`.

**Reply by Ghostbird** [reply, marked as answer]
It must've been something related to the specific type I was trying to generate. I called the code above on another type, and it worked out of the box. I have no idea why.
